src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+    });
+
+    map.on('mousemove', function(evt) {
+        var bb = self.div.getBoundingClientRect();
+
+        // some hackery to get Fx.hover to work
+
+        evt.clientX = evt.point.x + bb.left;
+        evt.clientY = evt.point.y + bb.top;
+
+        evt.target.getBoundingClientRect = function() { return bb; };
+
+        self.xaxis.p2c = function() { return evt.lngLat.lng; };
+        self.yaxis.p2c = function() { return evt.lngLat.lat; };
+
+        Fx.hover(gd, evt, self.id);
+    });
+
+    map.on('click', function() {
+        Fx.click(gd, { target: true });
+    });
+
+    function unhover() {
+        Fx.loneUnhover(fullLayout._toppaper);
+    }
+
+    map.on('dragstart', unhover);
+    map.on('zoomstart', unhover);
+
+};
+
+proto.updateMap = function(calcData, fullLayout, resolve, reject) {
+    var self = this,
+        map = self.map;
+
+    self.rejectOnError(reject);
+
+    var styleUrl = convertStyleUrl(self.opts.style);
+
+    if(self.styleUrl !== styleUrl) {
+        self.styleUrl = styleUrl;
+        map.setStyle(styleUrl);
+
+        map.style.once('load', function() {
+
+            // need to rebuild trace layers on reload
+            // to avoid 'lost event' errors
+            self.traceHash = {};
+
+            self.updateData(calcData);
+            self.updateLayout(fullLayout);
+
+            self.resolveOnRender(resolve);
+        });
+    }
+    else {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    }
+};
+
+proto.updateData = function(calcData) {
+    var traceHash = this.traceHash;
+
+    var traceObj, trace, i, j;
+
+    // update or create trace objects
+    for(i = 0; i < calcData.length; i++) {
+        var calcTrace = calcData[i];
+
+        trace = calcTrace[0].trace;
+        traceObj = traceHash[trace.uid];
+
+        if(traceObj) traceObj.update(calcTrace);
+        else {
+            traceHash[trace.uid] = trace._module.plot(this, calcTrace);
+        }
+    }
+
+    // remove empty trace objects
+    var ids = Object.keys(traceHash);
+    id_loop:
+    for(i = 0; i < ids.length; i++) {
+        var id = ids[i];
+
+        for(j = 0; j < calcData.length; j++) {
+            trace = calcData[j][0].trace;
+
+            if(id === trace.uid) continue id_loop;
+        }
+
+        traceObj = traceHash[id];
+        traceObj.dispose();
+        delete traceHash[id];
+    }
+};
+
+proto.updateLayout = function(fullLayout) {
+    var map = this.map,
+        opts = this.opts;
+
+    map.setCenter(convertCenter(opts.center));
+    map.setZoom(opts.zoom);
+    map.setBearing(opts.bearing);
+    map.setPitch(opts.pitch);
+
+    this.updateLayers();
+    this.updateFramework(fullLayout);
+    this.map.resize();
+};
+
+proto.resolveOnRender = function(resolve) {
+    var map = this.map;
+
+    map.on('render', function onRender() {
+        if(map.loaded()) {
+            map.off('render', onRender);
```

Comment:
believe it or not, this is recommended way of doing this ([ref](https://github.com/mapbox/mapbox-gl-js/issues/1715#issuecomment-189046453))